### PR TITLE
Initial Purification Optimization

### PR DESCRIPTION
**Pull request by tillarnold**
@vakaras this is some initial code for the Purification Optimization. The way that the bodies of the automatized functions is generated here is wrong I'm pretty sure.

### CODE REVIEW

File: prusti-viper/src/encoder/encoder.rs [hunk 1]
```diff
@@ -274,10 +276,103 @@ impl<'v, 'tcx> Encoder<'v, 'tcx> {
...
+        .borrow()
+        .clone();
+
+        for f in self.pure_functions.borrow().values().into_iter() {
+            let old_formal_args = f.formal_args.clone();
+            let mut purifier = ExprPurifier {old_formal_args: old_formal_args.clone(), snapshots: snapshots_info.clone()};
+
+            let mut formal_args: Vec<vir::LocalVar> = f.formal_args.iter().map(|e| {
+                    let old_type = e.typ.clone();
+                    let new_type =  purifier.transalte_type(old_type);
```
Comment (vakaras) [inline prusti-viper/src/encoder/encoder.rs]:
```suggestion
                    let new_type =  purifier.translate_type(old_type);
```

File: prusti-viper/src/encoder/encoder.rs [hunk 2]
```diff
@@ -1551,3 +1646,104 @@ impl<'v, 'tcx> Encoder<'v, 'tcx> {
         substs.join(";")
     }
 }
+
+
+
+struct ExprPurifier {
```
Comment (vakaras) [inline prusti-viper/src/encoder/encoder.rs]: Could you please move this to a separate file?

File: prusti-viper/src/encoder/snapshot_encoder.rs [hunk 3]
```diff
@@ -32,20 +33,21 @@ pub struct SnapshotEncoder<'p, 'v: 'p, 'tcx: 'v> {
...
 
-#[derive(Clone)]
+#[derive(Clone, Debug)]
 pub struct SnapshotDomain {
     pub domain: vir::Domain,
     pub equals_func: vir::Function,
     pub equals_func_ref: vir::Function,
     pub not_equals_func: vir::Function,
     pub not_equals_func_ref: vir::Function,
+    pub fields: Vec<String>,
```
Comment (vakaras) [inline prusti-viper/src/encoder/snapshot_encoder.rs]: Why do you need this?

File: prusti-common/src/vir/ast/expr.rs [hunk 4]
```diff
@@ -425,6 +430,13 @@ impl Expr {
         Expr::Cond(box guard, box left, box right, Position::default())
     }
 
+    pub fn int(value: i64) -> Self {
+        Expr::Const(
+            Const::Int(value),
+            Position::default(),
+        )
+    }
+
```
Comment (fpoli) [inline prusti-common/src/vir/ast/expr.rs:439]: We already have `Expr::from(value)`, which is implemented for many integer types.

File: prusti-viper/src/encoder/snapshot/purifier.rs [hunk 5]
```diff
@@ -56,16 +86,15 @@ impl ExprFolder for ExprPurifier {
     fn fold_local(&mut self, v: LocalVar, p: Position) -> Expr {
         if v.name == "__result" {
             self.self_function.clone()
+        } else {
+            Expr::Local(
+                LocalVar {
+                    name: v.name,
+                    typ: super::transalte_type(v.typ, &self.snapshots),
```
Comment (Aurel300) [inline prusti-viper/src/encoder/snapshot/purifier.rs]:
```suggestion
                    typ: super::translate_type(v.typ, &self.snapshots),
```

File: prusti-viper/src/encoder/snapshot_encoder.rs [hunk 6]
```diff
@@ -223,25 +245,31 @@ impl<'p, 'v, 'r: 'v, 'a: 'r, 'tcx: 'a> SnapshotEncoder<'p, 'v, 'tcx> {
...
                         }
                     }
                 }
 
                 true
             }
+            ty::TyKind::Adt(adt_def, subst) if adt_def.is_box() => {
+                parent_boxes.push(ty);
+                let is = self.is_ty_supported(ty.boxed_ty(), parent_boxes);
+                parent_boxes.retain(|&x| x != ty);
```
Comment (Aurel300) [inline prusti-viper/src/encoder/snapshot_encoder.rs]: Is it not enough to just `pop` here?

File: prusti-viper/src/encoder/snapshot_encoder.rs [hunk 7]
```diff
@@ -630,10 +658,56 @@ impl<'p, 'v, 'r: 'v, 'a: 'r, 'tcx: 'a> SnapshotEncoder<'p, 'v, 'tcx> {
...
     {
-        let snapshot = self.encoder.encode_snapshot(field_ty)?;
-        Ok(snapshot.snap_call(
-            self.encode_arg_field(location, field)
-        ))
+        warn!("encode_snap_arg begin");
+        let res = match field_ty.kind() {
+        ty::TyKind::Adt(adt_def, _) if adt_def.is_box() => { //TODO this breaks for lots of boxes. e.g. Box<Box<T>> or Box<i32>
+            let boxed_ty = field_ty.boxed_ty();
+            warn!("enocding boxed ty {:?}", boxed_ty);
```
Comment (Aurel300) [inline prusti-viper/src/encoder/snapshot_encoder.rs]: Can you invoke the snapshot encoding recursively here, i.e. `self.encoder.encode_snapshot(boxed_ty)`?

File: prusti-viper/src/encoder/encoder.rs [hunk 8]
```diff
@@ -249,10 +259,248 @@ impl<'v, 'tcx> Encoder<'v, 'tcx> {
...
+            domains.push(self.axiomatized_function_domain.borrow().clone());
+            domains.push(self.get_nat_domain());
+            domains.push(self.get_primitive_valid_domain())
+        }
+
         domains.sort_by_key(|d| d.get_identifier());
         domains
     }
 
+    fn get_primitive_valid_domain(&self) -> vir::Domain {
```
Comment (fpoli) [inline prusti-viper/src/encoder/encoder.rs]: This and the `get_nat_domain` implementation might go to `prusti-viper/src/encoder/builtin_encoder.rs` (e.g. adding a `BuiltinDomainKind`) to avoid making `encoder.rs` longer.

File: prusti-viper/src/encoder/snapshot_encoder.rs [hunk 9]
```diff
@@ -177,33 +180,61 @@ impl<'p, 'v, 'r: 'v, 'a: 'r, 'tcx: 'a> SnapshotEncoder<'p, 'v, 'tcx> {
...
+                            ).encode_snapshot()?
+                        } else {
+                            unreachable!()
+                        }
+                    }
+                    x => unimplemented!("{:?}", x),
+                }
             }
-
             x => unimplemented!("{:?}", x),
```
Comment (fpoli) [inline prusti-viper/src/encoder/snapshot_encoder.rs]: It would be better to return `EncodingError::unsupported(..)` instead of calling `unimplemented!(..)`.

File: prusti-viper/src/encoder/snapshot_encoder.rs [hunk 10]
```diff
@@ -555,58 +630,84 @@ impl<'p, 'v, 'r: 'v, 'a: 'r, 'tcx: 'a> SnapshotEncoder<'p, 'v, 'tcx> {
...
+                            vir::Position::default(),
+                        ))
+
+                        
+                    } else {
+                        unreachable!()
+                    }
                 }
-                args
+                x => unimplemented!("{:?}", x),
```
Comment (fpoli) [inline prusti-viper/src/encoder/snapshot_encoder.rs]: `EncodingError::unsupported(..)`?

File: prusti-viper/src/encoder/snapshot_encoder.rs [hunk 11]
```diff
@@ -647,24 +748,700 @@ impl<'p, 'v, 'r: 'v, 'a: 'r, 'tcx: 'a> SnapshotEncoder<'p, 'v, 'tcx> {
...
+                    let variant_name = variant.ident.name.to_ident_string();
+                    axioms.push(self.encode_enum_variant_cons_axiom(variant)?);
+                    for field in &variant.fields {
+                        funcs.push(self.encode_field_func(field, Some(variant_name.clone()))?);
+                        axioms.push(self.encode_field_axiom(field, Some(variant))?);
+                    }
+                }
+            }
+            _ => {
+                unimplemented!();
```
Comment (fpoli) [inline prusti-viper/src/encoder/snapshot_encoder.rs]: `EncodingError::unsupported(..)`?

File: prusti-viper/src/encoder/encoder.rs [hunk 12]
```diff
@@ -249,10 +259,248 @@ impl<'v, 'tcx> Encoder<'v, 'tcx> {
...
+        vir::Domain {
+            name: nat_domain_name.to_owned(),
+            functions,
+            axioms: vec![],
+            type_vars: vec![],
+        }
+    }
+
+    fn encode_axiomatized_pure_function(&self, f: &vir::Function) {
+        let snapshots_info: HashMap<String, Box<Snapshot>> = self.snapshots.borrow().clone();
```
Comment (fpoli) [inline prusti-viper/src/encoder/encoder.rs]: I wonder if it's possible to avoid cloning the map. The `ExprPurifier` could take a reference to the map, for example.

File: prusti-viper/src/encoder/snapshot_encoder.rs (hunk 7 above)
Comment (tillarnold) [reply prusti-viper/src/encoder/snapshot_encoder.rs]: I think that would cause infinite recursion. For example if I have a struct
```rust
struct List {
    value: u32,
    next: Option<Box<List>>,
}
```
`encode_snapshot(List)` would call `encode_snapshot(Option<Box<List>>)` and then that would call `encode_snapshot(Box<List>)` which calls `encode_snap_arg(Box<List>)` and then `encode_snapshot(List)` (because we'd call `self.encoder.encode_snapshot(boxed_ty)` where the `boxed_ty` is `List` )

This admittedly rather ugly code I wrote is an attempt to prevent this infinite recursion by only computing the parts of the snapshot we actually need.

Comment (Aurel300) [reply prusti-viper/src/encoder/snapshot_encoder.rs]: I understand the issue, but I just doing another (less complete) copy of the encoding for boxes is not really a solution. You could split ADT encoding into two phases, such that the `snapshot` map in `Encoder` contains an entry for the snapshot being encoded before the recursive case is hit.

File: prusti-viper/src/encoder/encoder.rs (hunk 12 above)
Comment (tillarnold) [reply prusti-viper/src/encoder/encoder.rs]: That was in fact possible. Thank you for pointing this out.

File: prusti-viper/src/encoder/encoder.rs (hunk 8 above)
Comment (tillarnold) [reply prusti-viper/src/encoder/encoder.rs]: As you suggested I have moved this to `builtin_encoder.rs` and added `BuiltinDomainKind`

**Comment by vakaras** [comment]
The changes of this PR were merged as part of #354.

File: prusti-viper/src/encoder/snapshot_encoder.rs [hunk 13]
```diff
@@ -647,24 +737,694 @@ impl<'p, 'v, 'r: 'v, 'a: 'r, 'tcx: 'a> SnapshotEncoder<'p, 'v, 'tcx> {
...
+                self.subst
+            )?;
+            Ok(
+                vir::Expr::unfolding(
+                    predicate_name,
+                    vec![variant_location],
+                    vir::Expr::domain_func_app(cons_func, args),
+                    vir::PermAmount::Read,
+                    enum_variant_index,
+                )
```
Comment (Aurel300) [inline prusti-viper/src/encoder/snapshot_encoder.rs:1392]: @tillarnold Why is this needed? Does the fold/unfold step fail to generate the variant unfolding automatically?

Comment (tillarnold) [reply prusti-viper/src/encoder/snapshot_encoder.rs:1392]: I am not sure why this is needed. I think (if I did git blame correctly) this was originally added in commit https://github.com/viperproject/prusti-dev/commit/8e0bf653d7e114e25e7e7f36dbc16f9c497983b0 so I don't think I'm the original author of this.
